File: ios/Sources/Drengr/Classify.swift
import Foundation

/// Seal-by-default body splitter (port of classify.ts). One pass over a captured
/// body splits every leaf into DROP (credentials/PCI -> [REDACTED-*], never stored),
/// SEAL (PII/unknown free-text -> raw into piiMap, typed placeholder into projection),
/// or KEEP (business signal: numbers/bools whose NAME doesn't read as PII, single-token
/// allowlisted enums -> plaintext). Typed placeholders keep leaf type stable.
/// GUARANTEE (honest scope): every free-text/spaced string, and every value matching a
/// PII PATTERN, seals by default; a value whose NAME hints PII seals even as a number.
/// The one residual is a BARE NUMBER under a name that gives no hint it is PII (a raw
/// coordinate named "x") — genuinely ambiguous, best-effort, not "never".
/// piiMap is collected for the (future) encrypt layer; only `projection` ships today.
enum Classify {
    struct Classified {
        let projection: String?
        let piiMap: [String: String]
        let piiPaths: [String]
    }

    private static let maxKeys = 512
    private static let maxDepth = 12
    private static let maxStr = 1024

    // DROP — pure secrets + PCI card data.
    private static let credentialNames: Set<String> = [
        "password", "passwd", "pwd", "pass", "passphrase", "secret", "clientsecret", "token",
        "apikey", "apisecret", "accesstoken", "refreshtoken", "idtoken", "oauthtoken", "bearertoken",
        "privatekey", "secretkey", "sessiontoken", "authorization", "auth", "otp", "totp",
        "csrf", "xsrf", "csrftoken", "xsrftoken",
        "cvv", "cvc", "cvv2", "csc", "pin", "cardnumber", "cardno", "ccnumber", "creditcard", "pan",
    ]

    // SEAL — PII the tenant may want decrypted, but Drengr must not.
    private static let piiNames: Set<String> = [
        "email", "phone", "mobile", "tel", "telephone", "fax",
        "firstname", "lastname", "middlename", "fullname", "username", "nickname",
        "customername", "recipientname", "sendername", "contactname",
        "ssn", "sin", "iban", "accountnumber", "routingnumber", "sortcode",
        "passport", "nationality", "dob", "dateofbirth", "birthdate",
        "address", "street", "zip", "zipcode", "postal", "postalcode",
        "lat", "latitude", "lng", "lon", "longitude", "geo", "coordinates",
        "ip", "ipaddress", "deviceid", "idfa", "gaid", "adid", "imei", "macaddress",
        "promocode", "promotioncode", "coupon", "giftmessage", "messagetext",
    ]

    // KEEP — business signal the discovery engine names events from.
    private static let businessAllowlist: Set<String> = [
        "status", "statuscode", "httpstatus", "responsecode", "code", "state", "result", "outcome",
        "declinereason", "declinecode", "reason", "errorcode",
        "currency", "amount", "price", "total", "subtotal", "tax", "shipping", "discount", "fee",
        "balance", "cost", "revenue", "quantity", "qty", "count",
        "sku", "productid", "itemid", "variantid", "orderid", "transactionid", "paymentid", "invoiceid",
        "plan", "planid", "tier", "type", "kind", "category", "subcategory",
        "event", "eventname", "action", "method", "httpmethod", "verb",
        "success", "ok", "enabled", "active", "error",
        "duration", "latency", "elapsed", "level", "score", "rating", "stars",
        "version", "appversion", "build", "step", "index", "page", "pagesize", "limit", "offset",
    ]

    // --- value detectors ---
    // The credential/PII value patterns + checksum primitives live in Patterns.swift,
    // shared with Redact so a hardening fix can't land in one file and miss the other.
    // credRes uses pemDetectRe (BEGIN prefix only) so a TRUNCATED private key still DROPs.
    private static let credRes: [NSRegularExpression] = [
        Patterns.jwtRe, Patterns.bearerRe, Patterns.secretTokenRe, Patterns.pemDetectRe,
    ]
    // Any whitespace at all — the single-token KEEP gate. A spaced/prose value under a
    // business-allowlisted name ("Jane Doe") is free-text PII, not an enum, so it seals.
    // Classify-only (not a PII value detector), so it stays here.
    private static let whitespaceRe = try! NSRegularExpression(pattern: "\\s")
    // Long, low-false-positive PII name fragments — a field whose NAME contains one seals
    // even a NUMERIC value (a national/tax id or geo coord sent as a number would otherwise
    // KEEP as a measure). Only long unambiguous tokens — never 'lat'/'ip' (→ 'latency'/'flip').
    private static let piiNameFragments: [String] = [
        "latitude", "longitude", "nationalid", "taxid", "passport", "driverlicense", "driverslicense",
        "socialsecurity", "accountnumber", "routingnumber", "creditcard", "cardnumber", "dateofbirth",
    ]
    private static func nameLooksPii(_ n: String) -> Bool {
        piiNameFragments.contains { n.contains($0) }
    }

    private static func matches(_ re: NSRegularExpression, _ s: String) -> Bool {
        re.firstMatch(in: s, range: NSRange(location: 0, length: (s as NSString).length)) != nil
    }

    private static func firstMatchString(_ re: NSRegularExpression, _ s: String) -> String? {
        let ns = s as NSString
        guard let m = re.firstMatch(in: s, range: NSRange(location: 0, length: ns.length)) else { return nil }
        return ns.substring(with: m.range)
    }

    private static func isBool(_ n: NSNumber) -> Bool {
        CFGetTypeID(n) == CFBooleanGetTypeID()
    }

    private static func credentialValue(_ s: String) -> Bool {
        credRes.contains { matches($0, s) }
    }

    private static func panValue(_ s: String) -> Bool {
        let ns = s as NSString
        guard let m = Patterns.digitRunRe.firstMatch(in: s, range: NSRange(location: 0, length: ns.length)) else { return false }
        let d = ns.substring(with: m.range).replacingOccurrences(of: "[ -]", with: "", options: .regularExpression)
        return Patterns.looksLikePan(d)
    }

    private static func piiKind(_ s: String) -> String? {
        if matches(Patterns.emailRe, s) { return "email" }
        if matches(Patterns.ssnRe, s) { return "ssn" }
        if matches(Patterns.phoneRe, s) || matches(Patterns.e164Re, s) { return "phone" }
        if matches(Patterns.macRe, s) { return "macaddress" } // before ip: the loose ipv6 regex also matches a MAC
        if matches(Patterns.ipv4Re, s) || matches(Patterns.ipv6Re, s) { return "ip" }
        if matches(Patterns.uuidRe, s) { return "deviceid" }
        if let ib = firstMatchString(Patterns.ibanRe, s), Patterns.isIbanValue(ib) { return "iban" }
        if matches(Patterns.cryptoRe, s) { return "wallet" }
        return nil
    }

    private static func sameTyped(_ v: Any, _ label: String) -> Any {
        if let n = v as? NSNumber { return isBool(n) ? false : 0 }
        return label
    }

    private static func jsonRaw(_ v: Any) -> String {
        if let d = try? JSONSerialization.data(withJSONObject: [v], options: []),
           let s = String(data: d, encoding: .utf8), s.count >= 2 {
            return String(s.dropFirst().dropLast())
        }
        return "null"
    }

    private enum Disp {
        case keep(Any)
        case drop(Any)
        case seal(Any, String, String)
    }

    private static func classifyLeaf(_ key: String, _ path: String, _ v: Any) -> Disp {
        let n = Patterns.norm(key)
        // 1. hard credentials (value or name) -> DROP (never sealed, never recoverable).
        if let s = v as? String, credentialValue(s) { return .drop(sameTyped(v, "[REDACTED-SECRET]")) }
        if credentialNames.contains(n) { return .drop(sameTyped(v, "[REDACTED-SECRET]")) }

        // 1b. oversized free text -> DROP before any value regex runs on it. The value
        //     patterns are bounded, but a huge leaf is never business signal, and this
        //     keeps the classifier's own passes cheap on a pathological body.
        if let s = v as? String, s.count > maxStr { return .drop("[FREETEXT:len=\(s.count)]") }

        // 2. a recognized PII KIND by value -> SEAL. Runs before the generic PAN drop
        //    so a recoverable-PII value (e.g. an IBAN) isn't mis-dropped as a card by a
        //    coincidental Luhn substring — a real card is not a piiKind, so it still drops below.
        if let s = v as? String, let vk = piiKind(s) { return .seal("[PII:\(vk)]", jsonRaw(v), path) }

        // 3. card-shaped digit run -> DROP (PCI: worthless as analytics, dangerous even sealed).
        if let s = v as? String, panValue(s) { return .drop(sameTyped(v, "[REDACTED-SECRET]")) }

        // 4. PII by NAME -> SEAL. Exact PII name, OR a long PII name-fragment so a national/
        //    tax id or geo coordinate sent as a NUMBER seals instead of KEEPing as a measure.
        if piiNames.contains(n) || nameLooksPii(n) { return .seal(sameTyped(v, "[PII:\(n)]"), jsonRaw(v), path) }

        // 5. numbers / booleans that survived credential+PII checks are business signal -> KEEP
        if v is NSNumber { return .keep(v) }

        // 6. strings: an allowlisted SINGLE-TOKEN enum -> KEEP; everything else seals. The
        //    no-whitespace gate closes the plaintext-name leak — a spaced/prose value
        //    ("Jane Doe", "call John back") under a business name now seals, not KEEPs.
        if let s = v as? String {
            if s.isEmpty { return .keep(s) }
            if businessAllowlist.contains(n), s.count <= 64, !matches(whitespaceRe, s) { return .keep(s) }
            return .seal("[PII]", jsonRaw(v), path)
        }
        return .drop(sameTyped(v, "[REDACTED]"))
    }

    static func classifyBody(_ body: String?) -> Classified {
        let empty = Classified(projection: nil, piiMap: [:], piiPaths: [])
        guard let body = body, !body.isEmpty else { return empty }
        guard let decoded = parseJson(body) ?? parseForm(body) else { return empty }

        var proj = [String: Any]()
        var piiMap = [String: String]()
        var piiPaths = [String]()
        walk("", "", decoded, 0, &proj, &piiMap, &piiPaths)

        if proj.isEmpty { return empty }
        guard let d = try? JSONSerialization.data(withJSONObject: proj, options: [.sortedKeys]),
              let s = String(data: d, encoding: .utf8) else { return empty }
        return Classified(projection: s, piiMap: piiMap, piiPaths: piiPaths)
    }

    private static func walk(_ prefix: String, _ key: String, _ v: Any, _ depth: Int,
                             _ proj: inout [String: Any], _ piiMap: inout [String: String],
                             _ piiPaths: inout [String]) {
        if proj.count >= maxKeys || depth > maxDepth { return }
        if let arr = v as? [Any] {
            var i = 0
            while i < arr.count && proj.count < maxKeys {
                walk(prefix.isEmpty ? "\(i)" : "\(prefix).\(i)", key, arr[i], depth + 1, &proj, &piiMap, &piiPaths)
                i += 1
            }
            return
        }
        if let dict = v as? [String: Any] {
            for (k, val) in dict where proj.count < maxKeys {
                walk(prefix.isEmpty ? k : "\(prefix).\(k)", k, val, depth + 1, &proj, &piiMap, &piiPaths)
            }
            return
        }
        if v is NSNull { return }
        switch classifyLeaf(key, prefix, v) {
        case .keep(let x): proj[prefix] = x
        case .drop(let x): proj[prefix] = x
        case .seal(let ph, let raw, let path):
            proj[prefix] = ph
            piiMap[path] = raw
            piiPaths.append(path)
        }
    }

    private static func parseJson(_ body: String) -> Any? {
        let t = body.trimmingCharacters(in: .whitespacesAndNewlines)
        guard t.hasPrefix("{") || t.hasPrefix("["), let data = body.data(using: .utf8) else { return nil }
        return try? JSONSerialization.jsonObject(with: data)
    }

    private static func parseForm(_ body: String) -> Any? {
        if !body.contains("=") || body.contains("\n") || body.contains(" ") { return nil }
        if body.range(of: "^[^=&]+=[^&]*(?:&[^=&]+=[^&]*)*$", options: .regularExpression) == nil { return nil }
        var map = [String: Any]()
        for pairSub in body.split(separator: "&", omittingEmptySubsequences: false) {
            let pair = String(pairSub)
            guard let eq = pair.firstIndex(of: "="), eq > pair.startIndex else { continue }
            let k = String(pair[..<eq])
            let v = String(pair[pair.index(after: eq)...])
            let key = k.replacingOccurrences(of: "+", with: " ").removingPercentEncoding ?? k
            let val = v.replacingOccurrences(of: "+", with: " ").removingPercentEncoding ?? v
            map[key] = val
        }
        return map.isEmpty ? nil : map
    }
}

File: android/drengr/src/main/kotlin/dev/drengr/sdk/Classify.kt
package dev.drengr.sdk

import org.json.JSONArray
import org.json.JSONObject

// Seal-by-default body splitter (port of classify.ts). One pass over a body splits
// every leaf into DROP (credentials/PCI → [REDACTED-*], never stored), SEAL (PII +
// unknown free-text → typed placeholder in the plaintext projection; raw goes to
// piiMap for on-device encryption), or KEEP (business scalars/allowlisted enums).
// Fail-closed: a value reaches the projection ONLY if credential-free AND
// (numeric/bool not-PII OR business-allowlisted). Typed placeholders keep leaf type.
internal object Classify {
    data class Result(val projection: String?, val piiMap: Map<String, String>, val piiPaths: List<String>)

    private const val MAX_KEYS = 512
    private const val MAX_DEPTH = 12
    private const val MAX_STR = 1024
    private val normRe = Regex("[_\\-$@.\\s]")
    private fun norm(name: String) = normRe.replace(name.lowercase(), "")

    private val CREDENTIAL_NAMES = setOf(
        "password", "passwd", "pwd", "pass", "passphrase", "secret", "clientsecret", "token",
        "apikey", "apisecret", "accesstoken", "refreshtoken", "idtoken", "oauthtoken", "bearertoken",
        "privatekey", "secretkey", "sessiontoken", "authorization", "auth", "otp", "totp",
        "csrf", "xsrf", "csrftoken", "xsrftoken",
        "cvv", "cvc", "cvv2", "csc", "pin", "cardnumber", "cardno", "ccnumber", "creditcard", "pan",
    )

    private val PII_NAMES = setOf(
        "email", "phone", "mobile", "tel", "telephone", "fax",
        "firstname", "lastname", "middlename", "fullname", "username", "nickname",
        "customername", "recipientname", "sendername", "contactname",
        "ssn", "sin", "iban", "accountnumber", "routingnumber", "sortcode",
        "passport", "nationality", "dob", "dateofbirth", "birthdate",
        "address", "street", "zip", "zipcode", "postal", "postalcode",
        "lat", "latitude", "lng", "lon", "longitude", "geo", "coordinates",
        "ip", "ipaddress", "deviceid", "idfa", "gaid", "adid", "imei", "macaddress",
        "promocode", "promotioncode", "coupon", "giftmessage", "messagetext",
    )

    private val BUSINESS_ALLOWLIST = setOf(
        "status", "statuscode", "httpstatus", "responsecode", "code", "state", "result", "outcome",
        "declinereason", "declinecode", "reason", "errorcode",
        "currency", "amount", "price", "total", "subtotal", "tax", "shipping", "discount", "fee",
        "balance", "cost", "revenue", "quantity", "qty", "count",
        "sku", "productid", "itemid", "variantid", "orderid", "transactionid", "paymentid", "invoiceid",
        "plan", "planid", "tier", "type", "kind", "category", "subcategory",
        "event", "eventname", "action", "method", "httpmethod", "verb",
        "success", "ok", "enabled", "active", "error",
        "duration", "latency", "elapsed", "level", "score", "rating", "stars",
        "version", "appversion", "build", "step", "index", "page", "pagesize", "limit", "offset",
    )

    private val credRe = listOf(
        Regex("eyJ[A-Za-z0-9_-]{6,}\\.[A-Za-z0-9_-]{6,}\\.[A-Za-z0-9_-]*"),
        Regex("[Bb]earer\\s+[A-Za-z0-9\\-._~+/]+=*"),
        Regex("\\b(?:(?:sk|rk)_(?:live|test)_[A-Za-z0-9]{16,}|AKIA[0-9A-Z]{16}|AIza[0-9A-Za-z_-]{35}|gh[opusr]_[A-Za-z0-9]{36,}|xox[baprs]-[A-Za-z0-9-]{10,})\\b"),
        Regex("-----BEGIN[A-Z0-9 ]*PRIVATE KEY-----"),
    )
    private val emailRe = Regex("[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}")
    private val ssnRe = Regex("\\b\\d{3}-\\d{2}-\\d{4}\\b")
    private val phoneRe = Regex("(?:\\+\\d{1,3}[ .-]?)?\\(?\\d{3}\\)?[ .-]\\d{3}[ .-]\\d{4}\\b")
    private val ipv4Re = Regex("\\b(?:(?:25[0-5]|2[0-4]\\d|1?\\d?\\d)\\.){3}(?:25[0-5]|2[0-4]\\d|1?\\d?\\d)\\b")
    private val ipv6Re = Regex("\\b(?:[A-Fa-f0-9]{1,4}:){2,7}[A-Fa-f0-9]{0,4}\\b")
    private val uuidRe = Regex("\\b[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\\b")
    private val digitRun = Regex("[0-9](?:[ -]?[0-9]){11,}")
    private val sep = Regex("[ -]")
    private val doubleSpace = Regex("\\s{2,}")

    private sealed class Disp
    private class Keep(val v: Any) : Disp()
    private class Drop(val v: Any) : Disp()
    private class Seal(val placeholder: Any, val raw: String, val path: String) : Disp()

    private fun luhn(digits: String): Boolean {
        if (digits.length < 13) return false
        var sum = 0
        var alt = false
        for (i in digits.length - 1 downTo 0) {
            var n = digits[i].code - 48
            if (n < 0 || n > 9) return false
            if (alt) { n *= 2; if (n > 9) n -= 9 }
            sum += n
            alt = !alt
        }
        return sum % 10 == 0
    }

    private fun credentialValue(s: String) = credRe.any { it.containsMatchIn(s) }

    private fun panValue(s: String): Boolean {
        val m = digitRun.find(s) ?: return false
        val d = sep.replace(m.value, "")
        if (d.length > 40) return true
        var len = 13
        while (len <= 19 && len <= d.length) {
            var i = 0
            while (i + len <= d.length) { if (luhn(d.substring(i, i + len))) return true; i++ }
            len++
        }
        return false
    }

    private fun piiKind(s: String): String? = when {
        emailRe.containsMatchIn(s) -> "email"
        ssnRe.containsMatchIn(s) -> "ssn"
        phoneRe.containsMatchIn(s) -> "phone"
        ipv4Re.containsMatchIn(s) || ipv6Re.containsMatchIn(s) -> "ip"
        uuidRe.containsMatchIn(s) -> "deviceid"
        else -> null
    }

    private fun sameTyped(v: Any, label: String): Any = when (v) {
        is Number -> 0
        is Boolean -> false
        else -> label
    }

    private fun jsonStringify(v: Any): String = if (v is String) JSONObject.quote(v) else v.toString()

    private fun classifyLeaf(key: String, path: String, v: Any): Disp {
        val n = norm(key)
        if (v is String && (credentialValue(v) || panValue(v))) return Drop(sameTyped(v, "[REDACTED-SECRET]"))
        if (n in CREDENTIAL_NAMES) return Drop(sameTyped(v, "[REDACTED-SECRET]"))

        val vk = if (v is String) piiKind(v) else null
        if (vk != null) return Seal("[PII:$vk]", jsonStringify(v), path)
        if (n in PII_NAMES) return Seal(sameTyped(v, "[PII:$n]"), jsonStringify(v), path)

        if (v is Number || v is Boolean) return Keep(v)

        if (v is String) {
            if (v.isEmpty()) return Keep(v)
            if (v.length > MAX_STR) return Drop("[FREETEXT:len=${v.length}]")
            if (n in BUSINESS_ALLOWLIST && v.length <= 64 && !doubleSpace.containsMatchIn(v)) return Keep(v)
            return Seal("[PII]", jsonStringify(v), path)
        }
        return Drop(sameTyped(v, "[REDACTED]"))
    }

    fun classifyBody(body: String?): Result {
        val empty = Result(null, emptyMap(), emptyList())
        if (body.isNullOrEmpty()) return empty
        val decoded = parseJson(body) ?: parseForm(body) ?: return empty

        val proj = JSONObject()
        val piiMap = LinkedHashMap<String, String>()
        val piiPaths = ArrayList<String>()

        fun walk(prefix: String, key: String, v: Any?, depth: Int) {
            if (proj.length() >= MAX_KEYS || depth > MAX_DEPTH) return
            when {
                v is JSONArray -> {
                    var i = 0
                    while (i < v.length() && proj.length() < MAX_KEYS) {
                        walk(if (prefix.isEmpty()) "$i" else "$prefix.$i", key, v.get(i), depth + 1); i++
                    }
                }
                v is JSONObject -> for (k in v.keys()) {
                    if (proj.length() >= MAX_KEYS) break
                    walk(if (prefix.isEmpty()) k else "$prefix.$k", k, v.get(k), depth + 1)
                }
                v == null || v === JSONObject.NULL -> {}
                else -> when (val d = classifyLeaf(key, prefix, v)) {
                    is Keep -> proj.put(prefix, d.v)
                    is Drop -> proj.put(prefix, d.v)
                    is Seal -> { proj.put(prefix, d.placeholder); piiMap[d.path] = d.raw; piiPaths.add(d.path) }
                }
            }
        }

        return try {
            walk("", "", decoded, 0)
            if (proj.length() == 0) empty else Result(proj.toString(), piiMap, piiPaths)
        } catch (_: Throwable) {
            empty
        }
    }

    private fun parseJson(body: String): Any? {
        val t = body.trimStart()
        if (t.isEmpty() || (t[0] != '{' && t[0] != '[')) return null
        return try {
            if (t[0] == '{') JSONObject(body) else JSONArray(body)
        } catch (_: Throwable) { null }
    }

    private val formShape = Regex("^[^=&]+=[^&]*(?:&[^=&]+=[^&]*)*$")
    private fun parseForm(body: String): JSONObject? {
        if (!body.contains('=') || body.contains('\n') || body.contains(' ')) return null
        if (!formShape.matches(body)) return null
        val o = JSONObject()
        for (pair in body.split('&')) {
            val i = pair.indexOf('='); if (i <= 0) continue
            val k = try { java.net.URLDecoder.decode(pair.substring(0, i), "UTF-8") } catch (_: Throwable) { pair.substring(0, i) }
            val v = try { java.net.URLDecoder.decode(pair.substring(i + 1), "UTF-8") } catch (_: Throwable) { pair.substring(i + 1) }
            o.put(k, v)
        }
        return if (o.length() == 0) null else o
    }
}
